rasa/core/policies/rule_policy.py

```diff
@@ -0,0 +1,281 @@
+from rasa.core.featurizers import TrackerFeaturizer
+from rasa.core.policies.memoization import MemoizationPolicy
+from rasa.core.policies.policy import SupportedData
+from rasa.core.trackers import DialogueStateTracker
+from rasa.core.constants import (
+    FORM_POLICY_PRIORITY,
+    RULE_SNIPPET_ACTION_NAME,
+    USER_INTENT_RESTART,
+    USER_INTENT_BACK,
+    USER_INTENT_SESSION_START,
+)
+from rasa.core.actions.action import (
+    ACTION_LISTEN_NAME,
+    ACTION_RESTART_NAME,
+    ACTION_BACK_NAME,
+    ACTION_SESSION_START_NAME,
+)
+
+logger = logging.getLogger(__name__)
+
+# These are Rasa Open Source default actions and overrule everything at any time.
+DEFAULT_ACTION_MAPPINGS = {
+    USER_INTENT_RESTART: ACTION_RESTART_NAME,
+    USER_INTENT_BACK: ACTION_BACK_NAME,
+    USER_INTENT_SESSION_START: ACTION_SESSION_START_NAME,
+}
+
+
+class RulePolicy(MemoizationPolicy):
+    """Policy which handles all the rules"""
+
+    ENABLE_FEATURE_STRING_COMPRESSION = False
+
+    @staticmethod
+    def supported_data() -> SupportedData:
+        """The type of data supported by this policy.
+
+        Returns:
+            The data type supported by this policy (rule data).
+        """
+        return SupportedData.RULE_DATA
+
+    def __init__(
+        self,
+        featurizer: Optional[TrackerFeaturizer] = None,
+        priority: int = FORM_POLICY_PRIORITY,
+        lookup: Optional[Dict] = None,
+    ) -> None:
+        if not featurizer:
+            # max history is set to `None` in order to capture lengths of rule stories
+            featurizer = self._standard_featurizer()
+            featurizer.max_history = None
+
+        super().__init__(featurizer=featurizer, priority=priority, lookup=lookup)
+
+    def _create_feature_key(self, states: List[Dict]):
+
+        feature_str = ""
+        for state in states:
+            if state:
+                feature_str += "|"
+                for feature in state.keys():
+                    feature_str += feature + " "
+                feature_str = feature_str.strip()
+
+        return feature_str
+
+    @staticmethod
+    def _features_in_state(fs, state):
+
+        state_slots = defaultdict(set)
+        for s in state.keys():
+            if s.startswith("slot"):
+                state_slots[s[: s.rfind("_")]].add(s)
+
+        f_slots = defaultdict(set)
+        for f in fs:
+            if f.endswith("_None"):
+                # TODO: this is a hack to make a rule know
+                #  that slot or form should not be set
+                if any(f[: f.rfind("_")] in key for key in state.keys()):
+                    return False
+            elif f not in state:
+                return False
+            elif f.startswith("slot"):
+                f_slots[f[: f.rfind("_")]].add(f)
+
+        for k, v in f_slots.items():
+            if state_slots[k] != v:
+                return False
+
+        return True
+
+    def _rule_is_good(self, key, i, state):
```

Comment:
types

Reply:
can we please update the variable names and add a comment when a rule is good? It takes you sometime to understand the below statement.

Reply:
@Ghostvv I added types but the comment would be great